core/src/oracles.ts

```diff
@@ -0,0 +1,145 @@
+         **/
+        const currentPriceFeed = await provider.getStorageAt(oracleAddress, oracle.currentPriceSlotAddress);
+        const slotPriceValueBeginsAtPosition = oracle.slotPriceValueBeginsAtPosition;
+        const priceValiditySlotValue = await provider.getStorageAt(
+            oracleAddress,
+            oracle.currentPriceValiditySlotAndOffset.slot
+        );
+        const isPriceValid =
+            parseInt(priceValiditySlotValue[oracle.currentPriceValiditySlotAndOffset.offset], 16) === 1;
+        return isPriceValid
+            ? new BigNumber(`0x${currentPriceFeed.substring(slotPriceValueBeginsAtPosition)}`)
+            : new BigNumber(NaN);
+    },
+    CurrentAndNextPrice: async (
+        oracle: OracleCurrentAndNextPrices,
+        provider: ethers.providers.JsonRpcProvider,
+        oracleAddress: string
+    ) => {
+        return await getOraclePriceSameSlotValidity(
+            oracle.currentPriceSlotAddress,
+            oracle.slotPriceValueBeginsAtPosition,
+            provider,
+            oracleAddress
+        );
+    },
+};
+const getCurrentOraclePrice = async (
+    oracle: CollateralPriceSourceConfig,
+    provider: ethers.providers.JsonRpcProvider,
+    oracleAddress: string
+) => {
+    return currentPriceExtractors[oracle.type](oracle, provider, oracleAddress);
+};
+
+const getNextOraclePriceChange = async (
+    collateralConfig: CollateralConfig,
+    provider: ethers.providers.JsonRpcProvider,
+    oracleAddress: string
+) => {
+    let nextPriceChange: Date = new Date(NaN);
+    if (collateralConfig.oracle.hasDelay) {
+        const osmContractInterface = await getContractInterfaceByName('OSM');
+        const osmContract = new ethers.Contract(oracleAddress, osmContractInterface, provider);
+        const lastPriceUpdateAsHex = (await osmContract.zzz())._hex;
+        const priceUpdateFrequencyInSeconds = await osmContract.hop();
+        const lastPriceUpdateTimestampInSeconds = new BigNumber(lastPriceUpdateAsHex).toNumber();
+        nextPriceChange = new Date((lastPriceUpdateTimestampInSeconds + priceUpdateFrequencyInSeconds) * 1000);
+    }
+    return nextPriceChange;
+};
+
+const _getOsmPrices = async (
+    network: string,
+    oracleAddress: string,
+    collateralType: CollateralType
+): Promise<OraclePrices> => {
+    const provider = await getProvider(network);
+    const collateralConfig = COLLATERALS[collateralType];
+
+    const nextPrice = await getNextOraclePrice(collateralConfig.oracle, provider, oracleAddress);
+    const currentPrice = await getCurrentOraclePrice(collateralConfig.oracle, provider, oracleAddress);
+    const nextPriceChange = await getNextOraclePriceChange(collateralConfig, provider, oracleAddress);
+
+    const currentUnitCollateralPrice = new BigNumber(currentPrice).shiftedBy(-WAD_NUMBER_OF_DIGITS);
+    const nextUnitCollateralPrice = new BigNumber(nextPrice).shiftedBy(-WAD_NUMBER_OF_DIGITS);
```

Comment:
- [ ] Instead of doing it here, please make sure that every function in our codebase returns tokens that are properly formatted already (otherwise it's a mess)
- [ ] Please always use correct unit, if you format DAI, use `DAI_NUMBER_OF_DIGITS`

Reply:
I see that you've returned those two lines, but I don't see that they've been moved to the respective functions. Where did they go?

Reply:
the `shiftedBy` now is executing right at the place where the prices are fetched (there're 2 places in this file), there're no additional variable assignments - it's just `return ...shiftedBy()`.